src/ciphersuite/signable.rs

```diff
@@ -1,29 +1,47 @@
 use crate::ciphersuite::*;
 use crate::credentials::*;
 
+/// This trait must be implemented by all structs that contain a self-signature.
+pub trait SignedStruct<T> {
+    /// Build a signed struct version from the payload struct.
+    fn from_payload(payload: T, signature: Signature) -> Self;
+}
+
 /// The `Signable` trait is implemented by all struct that are being signed.
 /// The implementation has to provide the `unsigned_payload` function.
 pub trait Signable: Sized {
+    type SignedOutput;
+
     fn unsigned_payload(&self) -> Result<Vec<u8>, crate::codec::CodecError>;
 
     /// Sign the payload with the given `id`.
     ///
     /// Returns a `Signature`.
-    fn sign(&self, credential_bundle: &CredentialBundle) -> Signature {
-        let payload = self.unsigned_payload().unwrap();
-        credential_bundle.sign(&payload).unwrap()
+    fn sign(
+        self,
+        credential_bundle: &CredentialBundle,
+    ) -> Result<Self::SignedOutput, CredentialError>
+    where
+        Self::SignedOutput: SignedStruct<Self>,
+    {
+        let payload = self.unsigned_payload()?;
+        let signature = credential_bundle
+            .sign(&payload)
+            .map_err(|_| CredentialError::SignatureError)?;
+        Ok(Self::SignedOutput::from_payload(self, signature))
     }
+}
+
+pub trait Verifiable {
```

Comment:
Maybe add some documentation on the idea behind this trait? You mention later on that Signable and Verifiable shouldn't be implemented by the same struct. Maybe that would be worth mentioning here.